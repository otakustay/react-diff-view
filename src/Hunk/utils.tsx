// @ts-nocheck ignore: third-party
import {computeOldLineNumber, computeNewLineNumber} from '../utils';

export const renderDefaultBy = (change, side) => () => {
    const lineNumber = side === 'old' ? computeOldLineNumber(change) : computeNewLineNumber(change);
    return lineNumber === -1 ? undefined : lineNumber;
};

// eslint-disable-next-line react/function-component-definition
export const wrapInAnchorBy = (gutterAnchor, anchorTarget) => element => {
    if (!gutterAnchor || !element) {
        return element;
    }

    return <a href={'#' + anchorTarget}>{element}</a>;
};

export const composeCallback = (own, custom) => {
    if (custom) {
        return e => {
            own(e);
            custom(); // `custom` is already bound with `arg`
        };
    }

    return own;
};
