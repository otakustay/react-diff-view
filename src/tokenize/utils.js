import {last} from 'lodash';

export const clone = path => path.map(node => ({...node}));

export const replace = (path, leaf) => [...clone(path.slice(0, -1)), leaf];

export const wrap = (path, parent) => [parent, ...clone(path)];

export const split = (path, splitStart, splitEnd, wrapSplitNode) => {
    const parents = path.slice(0, -1);
    const leaf = last(path);
    const output = [];

    if (splitEnd <= 0 || splitStart >= leaf.value.length) {
        return [path];
    }

    const split = (start, end) => {
        const value = leaf.value.slice(start, end);
        return [...parents, {...leaf, value}];
    };

    if (splitStart > 0) {
        const head = split(0, splitStart);
        output.push(clone(head));
    }

    const body = split(Math.max(splitStart, 0), splitEnd);
    output.push(wrapSplitNode ? wrap(body, wrapSplitNode) : clone(body));

    if (splitEnd < leaf.value.length) {
        const tail = split(splitEnd);
        output.push(clone(tail));
    }

    return output;
};
