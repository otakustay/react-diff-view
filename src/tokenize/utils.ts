import {ProcessingNode, TextNode, TokenPath} from './interface';

export function clone(path: TokenPath): TokenPath {
    return path.map(node => ({...node}));
}

export function replace(path: TokenPath, leaf: ProcessingNode): TokenPath {
    return [...clone(path.slice(0, -1)), leaf];
}

export function wrap(path: TokenPath, parent: ProcessingNode): TokenPath {
    return [parent, ...clone(path)];
}

function isTextNode(node: ProcessingNode): node is TextNode {
    return node.type === 'text';
}

export function leafOf(path: TokenPath): TextNode {
    const last = path[path.length - 1];

    if (isTextNode(last)) {
        return last;
    }

    throw new Error(`Invalid token path with leaf of type ${last.type}`);
}

export function split(path: TokenPath, splitStart: number, splitEnd: number, wrapSplitNode: ProcessingNode) {
    const parents = path.slice(0, -1);
    const leaf = leafOf(path);
    const output = [];

    if (splitEnd <= 0 || splitStart >= leaf?.value.length) {
        return [path];
    }

    const split = (start: number, end?: number) => {
        const value = leaf.value.slice(start, end);
        return [...parents, {...leaf, value}];
    };

    if (splitStart > 0) {
        const head = split(0, splitStart);
        output.push(clone(head));
    }

    const body = split(Math.max(splitStart, 0), splitEnd);
    output.push(wrapSplitNode ? wrap(body, wrapSplitNode) : clone(body));

    if (splitEnd < leaf.value.length) {
        const tail = split(splitEnd);
        output.push(clone(tail));
    }

    return output;
}
