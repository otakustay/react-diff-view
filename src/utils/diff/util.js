export const first = array => array[0];

export const last = array => array[array.length - 1];

export const sideToProperty = side => [side + 'Start', side + 'Lines'];
