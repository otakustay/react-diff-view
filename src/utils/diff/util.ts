export function first<T>(array: T[]) {
    return array[0];
}

export function last<T>(array: T[]) {
    return array[array.length - 1];
}

export function sideToProperty(side: 'old' | 'new') {
    return [`${side}Start`, `${side}Lines`] as const;
}
