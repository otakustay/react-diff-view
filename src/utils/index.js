export * from './diff';
export * from './callbacks';
export * from './parse';
