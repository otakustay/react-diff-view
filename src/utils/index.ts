export {
    computeNewLineNumber,
    computeOldLineNumber,
    expandCollapsedBlockBy,
    expandFromRawCode,
    findChangeByNewLineNumber,
    findChangeByOldLineNumber,
    getChangeKey,
    getCollapsedLinesCountBetween,
    getCorrespondingNewLineNumber,
    getCorrespondingOldLineNumber,
    insertHunk,
    textLinesToHunk,
} from './diff';
export {parseDiff} from './parse';
export type {Source} from './diff';
export type {ParseOptions, FileData, HunkData, ChangeData} from './parse';
