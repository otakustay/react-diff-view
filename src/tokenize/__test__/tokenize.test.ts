import {describe, test, expect} from 'vitest';
import refractor from 'refractor';
import dedent from 'dedent';
import {parseDiff} from '../../utils';
import {markEdits, pickRanges, tokenize, TokenizeOptions} from '../index';

describe('tokenize', () => {
    test('basic', () => {
        const expected = {
            new: [[{type: 'text', value: ''}]],
            old: [[{type: 'text', value: ''}]],
        };
        expect(tokenize([], {})).toEqual(expected);
    });

    test('highlight', () => {
        const result = tokenize(
            [],
            {
                highlight: true,
                refractor,
                language: 'javascript',
            }
        );
        expect(result).toMatchSnapshot();

    });

    test('with old source', () => {
        const expected = {
            new: [[{type: 'text', value: 'A'}]],
            old: [[{type: 'text', value: 'A'}]],
        };
        expect(tokenize([], {oldSource: 'A'})).toEqual(expected);
    });

    test('withHunk', () => {
        const diff = dedent`
            diff --git a/package.json b/package.json
            index 4778f48..c0edd5f 100644
            --- a/package.json
            +++ b/package.json
            @@ -13,7 +13,7 @@
                 ],
                 "scripts": {
                 "prepare": "husky install",
            -    "test": "jest --config ./jest.config.js",
            +    "test": "vitest run --coverage",
                 "start": "skr dev --src-dir=site",
                 "clean": "rm -rf cjs es style types",
                 "build": "sh scripts/build.sh",
            @@ -49,11 +49,12 @@
                 "@reskript/settings": "5.7.4",
                 "@rollup/plugin-typescript": "^11.0.0",
                 "@types/dedent": "^0.7.0",
            +    "@types/diff-match-patch": "^1.0.32",
                 "@types/lodash": "^4.14.191",
                 "@types/react-test-renderer": "^18.0.0",
            +    "@types/refractor": "^2.8.0",
                 "antd": "^5.2.1",
                 "autoprefixer": "^10.4.13",
            -    "babel-jest": "^29.4.3",
                 "babel-loader": "^9.1.2",
                 "babel-plugin-add-react-displayname": "0.0.5",
                 "babel-plugin-import": "^1.13.6",
            @@ -66,14 +67,12 @@
                 "enzyme": "^3.11.0",
                 "enzyme-adapter-react-16": "^1.15.7",
                 "eslint": "^8.34.0",
            -    "eslint-plugin-jest": "^27.2.1",
                 "eslint-plugin-react": "^7.32.2",
                 "eslint-plugin-react-hooks": "^4.6.0",
                 "gitdiff-parser": "^0.2.2",
                 "html-webpack-plugin": "^5.5.0",
                 "husky": "^8.0.3",
                 "identity-obj-proxy": "^3.0.0",
            -    "jest": "^29.4.3",
                 "less": "^4.1.3",
                 "less-loader": "^11.1.0",
                 "lodash": "^4.17.21",
            @@ -101,7 +100,6 @@
                 "sha1": "^1.1.1",
                 "style-loader": "^3.3.1",
                 "styled-components": "^5.3.6",
            -    "ts-jest": "^29.0.5",
                 "typescript": "^4.9.5",
                 "unidiff": "^1.0.2",
                 "vitest": "^0.28.5",
        `;
        const [file] = parseDiff(diff);
        expect(tokenize(file.hunks)).toMatchSnapshot();
    });

    test('enhance', () => {
        const oldSource = dedent`
            // Copyright
            package com.xxx;
            import java.util.List;
            import java.util.Set;
        `;

        const diffText = `
            diff --git a/a b/b
            index 5b25e5b..772d084 100644
            --- a/a
            +++ b/b
            @@ -2,6 +2,7 @@
             package com.xxx;
            +import java.util.Date;
             import java.util.List;
             import java.util.Set;
        `;

        const defs = [
            {
                id: 'yz5k9m0BvISpUQ2asedw',
                type: 'def',
                row: 2,
                col: 9,
                binding: 'package',
                length: 7,
                token: 'com.xxx',
            },
        ];

        const [file] = parseDiff(diffText, {nearbySequences: 'zip'});
        const options: TokenizeOptions = {
            oldSource,
            highlight: false,
            enhancers: [
                markEdits(file.hunks, {type: 'block'}),
                pickRanges([], defs.map(i => ({...i, lineNumber: i.row, start: i.col - 1}))),
            ],
        };
        const tokens = tokenize(file.hunks, options);
        expect(tokens).toMatchSnapshot();
    });
});
