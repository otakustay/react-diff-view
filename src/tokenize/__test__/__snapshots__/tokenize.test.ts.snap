// Vitest Snapshot v1

exports[`tokenize > enhance 1`] = `
{
  "new": [
    [
      {
        "type": "text",
        "value": "// Copyright",
      },
    ],
    [
      {
        "type": "text",
        "value": "package ",
      },
      {
        "binding": "package",
        "children": [
          {
            "type": "text",
            "value": "com.xxx",
          },
        ],
        "col": 9,
        "id": "yz5k9m0BvISpUQ2asedw",
        "length": 7,
        "lineNumber": 2,
        "row": 2,
        "start": 8,
        "token": "com.xxx",
        "type": "def",
      },
      {
        "type": "text",
        "value": ";",
      },
    ],
    [
      {
        "type": "text",
        "value": "import java.util.List;",
      },
    ],
    [
      {
        "type": "text",
        "value": "import java.util.Set;",
      },
    ],
  ],
  "old": [
    [
      {
        "type": "text",
        "value": "// Copyright",
      },
    ],
    [
      {
        "type": "text",
        "value": "package com.xxx;",
      },
    ],
    [
      {
        "type": "text",
        "value": "import java.util.List;",
      },
    ],
    [
      {
        "type": "text",
        "value": "import java.util.Set;",
      },
    ],
  ],
}
`;

exports[`tokenize > highlight 1`] = `
{
  "new": [
    [],
  ],
  "old": [
    [],
  ],
}
`;

exports[`tokenize > withHunk 1`] = `
{
  "new": [
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    ],",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"scripts\\": {",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"prepare\\": \\"husky install\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"test\\": \\"vitest run --coverage\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"start\\": \\"skr dev --src-dir=site\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"clean\\": \\"rm -rf cjs es style types\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"build\\": \\"sh scripts/build.sh\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@reskript/settings\\": \\"5.7.4\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@rollup/plugin-typescript\\": \\"^11.0.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/dedent\\": \\"^0.7.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/diff-match-patch\\": \\"^1.0.32\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/lodash\\": \\"^4.14.191\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/react-test-renderer\\": \\"^18.0.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/refractor\\": \\"^2.8.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"antd\\": \\"^5.2.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"autoprefixer\\": \\"^10.4.13\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-loader\\": \\"^9.1.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-plugin-add-react-displayname\\": \\"0.0.5\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-plugin-import\\": \\"^1.13.6\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"enzyme\\": \\"^3.11.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"enzyme-adapter-react-16\\": \\"^1.15.7\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint\\": \\"^8.34.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint-plugin-react\\": \\"^7.32.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint-plugin-react-hooks\\": \\"^4.6.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"gitdiff-parser\\": \\"^0.2.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"html-webpack-plugin\\": \\"^5.5.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"husky\\": \\"^8.0.3\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"identity-obj-proxy\\": \\"^3.0.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"less\\": \\"^4.1.3\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"less-loader\\": \\"^11.1.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"lodash\\": \\"^4.17.21\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"sha1\\": \\"^1.1.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"style-loader\\": \\"^3.3.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"styled-components\\": \\"^5.3.6\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"typescript\\": \\"^4.9.5\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"unidiff\\": \\"^1.0.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"vitest\\": \\"^0.28.5\\",",
      },
    ],
  ],
  "old": [
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    ],",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"scripts\\": {",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"prepare\\": \\"husky install\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"test\\": \\"jest --config ./jest.config.js\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"start\\": \\"skr dev --src-dir=site\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"clean\\": \\"rm -rf cjs es style types\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"build\\": \\"sh scripts/build.sh\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@reskript/settings\\": \\"5.7.4\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@rollup/plugin-typescript\\": \\"^11.0.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/dedent\\": \\"^0.7.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/lodash\\": \\"^4.14.191\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"@types/react-test-renderer\\": \\"^18.0.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"antd\\": \\"^5.2.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"autoprefixer\\": \\"^10.4.13\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-jest\\": \\"^29.4.3\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-loader\\": \\"^9.1.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-plugin-add-react-displayname\\": \\"0.0.5\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"babel-plugin-import\\": \\"^1.13.6\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"enzyme\\": \\"^3.11.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"enzyme-adapter-react-16\\": \\"^1.15.7\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint\\": \\"^8.34.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint-plugin-jest\\": \\"^27.2.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint-plugin-react\\": \\"^7.32.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"eslint-plugin-react-hooks\\": \\"^4.6.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"gitdiff-parser\\": \\"^0.2.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"html-webpack-plugin\\": \\"^5.5.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"husky\\": \\"^8.0.3\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"identity-obj-proxy\\": \\"^3.0.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"jest\\": \\"^29.4.3\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"less\\": \\"^4.1.3\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"less-loader\\": \\"^11.1.0\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"lodash\\": \\"^4.17.21\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"sha1\\": \\"^1.1.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"style-loader\\": \\"^3.3.1\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"styled-components\\": \\"^5.3.6\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"ts-jest\\": \\"^29.0.5\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"typescript\\": \\"^4.9.5\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"unidiff\\": \\"^1.0.2\\",",
      },
    ],
    [
      {
        "type": "text",
        "value": "    \\"vitest\\": \\"^0.28.5\\",",
      },
    ],
  ],
}
`;
